llvm/lib/CodeGen/MachineLICM.cpp

```diff
@@ -639,6 +639,21 @@ void MachineLICMImpl::HoistRegionPostRA(MachineLoop *CurLoop,
     if (const uint32_t *Mask = BB->getBeginClobberMask(TRI))
       applyBitsNotInRegMaskToRegUnitsMask(*TRI, RUClobbers, Mask);
 
+    // EH landing pads clobber exception pointer/selector registers
+    if (BB->isEHPad()) {
+      const MachineFunction &MF = *BB->getParent();
+      if (MF.getFunction().hasPersonalityFn()) {
+        auto PersonalityFn = MF.getFunction().getPersonalityFn();
+        const TargetLowering &TLI = *MF.getSubtarget().getTargetLowering();
+        if (unsigned Reg = TLI.getExceptionPointerRegister(PersonalityFn))
+          for (MCRegUnitIterator RUI(Reg, TRI); RUI.isValid(); ++RUI)
+            RUClobbers.set(*RUI);
+        if (unsigned Reg = TLI.getExceptionSelectorRegister(PersonalityFn))
+          for (MCRegUnitIterator RUI(Reg, TRI); RUI.isValid(); ++RUI)
+            RUClobbers.set(*RUI);
+      }
+    }
```

Comment:
This is the wrong place for this. This pass is already inventing its own liveness tracking instead of LiveRegUnits, which it should be using.

Why is this case special? Are these not reserved and/or callee saved?

Reply:
They are not reserved, but they are callee saved.  That's why they're considered candidates to be hoisted even though there's a call in the loop.  And normally that would be fine except for the special EH landing pad case.

I'm not sure where else these should be handled, I simply noticed that several other places also handle these registers specially.

Reply:
This should be just
```
    if (BB->isEHPad()) {
      if (const uint32_t *Mask =
              TRI->getCustomEHPadPreservedMask(*BB->getParent()))
        applyBitsNotInRegMaskToRegUnitsMask(*TRI, RUClobbers, Mask);
    }
```

Note however that this only partly addresses the issue. The `continue` at the beginning of the loop is wrong because it results in un-analyzed instructions. If an inner loop had a register def/clobber, it won't be taken into account by the loop over Candidates below.

I've been debugging a similar issue for quite some time now, but I'm struggling to reproduce it on an upstream target =\

Reply:
> Should this be just

Please ignore this. It appears this method only needs to be implemented if the unwinder clobbers registers other than exception pointer/selector. DWARF exception handling would usually preserve those other registers.

That said, it *should* be queried by LICM, but it is a different issue.

Reply:
I wonder whether `livein_iterator` (used just above) should take care of these registers implicitly like `liveout_iterator` does
https://github.com/llvm/llvm-project/blob/66f158d91803875de63d8f2a437ce8ecb22c4141/llvm/include/llvm/CodeGen/MachineBasicBlock.h#L550-L559

Reply:
Thanks, that makes sense to me now.